### ISSUE TITLE
False positive "Overloaded function signature 2 will never be matched: signature 1's parameter type(s) are the same or broader" when using Unions vs TypeVars

### ISSUE THREAD

**Issue by me-and** [python/mypy#15683]
**Bug Report**
When overloading a function to specify it can return `bytes` if given only `bytes` objects, `str` if given only `str` objects, or either if given a mix of both, mypy incorrectly complains that `AnyStr` is the same or broader than `Union[str, bytes]`.

Despite the warning, mypy _does_ correctly match the `Union[str, bytes]` signature when appropriate

**To Reproduce**

Very simple test case:

```python
from typing import overload, TypeAlias, Union, AnyStr, reveal_type
from collections.abc import Sequence
from random import choice

AllStr: TypeAlias = Union[bytes, str]

# Below produces "error: Overloaded function signature 2 will never be
# matched: signature 1's parameter type(s) are the same or broader
# [misc]"
@overload
def pick_three(choices: Sequence[AnyStr]) -> AnyStr: ...
@overload
def pick_three(choices: Sequence[AllStr]) -> AllStr: ...
def pick_three(choices: Sequence[AllStr]) -> AllStr:
    return choice(choices)

strings: tuple[str, ...] = ('abc', 'def')
byte_strings: tuple[bytes, ...] = (b'abc', b'def')
mixed_strings: tuple[AllStr, ...] = ('abc', b'def')

reveal_type(pick_three(strings))  # str
reveal_type(pick_three(byte_strings))  # bytes
reveal_type(pick_three(mixed_strings))  # Union[bytes, str], despite the warning
```

[Some more example code on mypy Playground, with other related cases that work as expected](https://mypy-play.net/?mypy=latest&python=3.11&gist=c84e3fe5d6f0c8a393d8b8aeb4dfcd58).

**Expected Behavior**

Mypy does not warn that signature 2 cannot match, given it clearly _is_ able to match based on the `reveal_type(pick_three(mixed_strings))` output

**Actual Behavior**
```
main.py:39: error: Overloaded function signature 2 will never be matched: signature 1's parameter type(s) are the same or broader  [misc]
main.py:43: note: Revealed type is "builtins.str"
main.py:44: note: Revealed type is "builtins.bytes"
main.py:45: note: Revealed type is "Union[builtins.bytes, builtins.str]"
```

**Your Environment**

Seen on Debian with Python 3.11 and mypy 1.0.1 using `--strict`, and in mypy Playground with Python 3.11 and mypy 1.4.1 with default options.
